miden-lib/asm/sat/internal/prologue.masm

```diff
@@ -133,6 +134,91 @@ end
 # ACCOUNT DATA
 # =================================================================================================
 
+#! Ingests the account storage slot type data and stores it in memory at the appropriate memory
+#! addresses.
+#!
+#! This is achieved by reading the storage type commitment from storage slot 255 and then reading
+#! the storage type data from the advice provider.
+#!
+#! Stack: []
+#! Merkle Store: { ACCT_STORAGE[slot_255] -> TYPES_COM }
+#! Advice Map: { TYPES_COM: [slot_0_type, slot_1_type, ..., slot_255_type] }
+#! Output: []
+proc.ingest_acct_storage_types
+    # get the storage slot types commitment
+    exec.account::get_slot_types_commitment_storage_slot exec.account::get_item
+    # => [TYPES_COM]
+
+    # load the type data from the advice map onto the advice stack
+    # we prefix with [push.1 drop] due to https://github.com/0xPolygonMiden/miden-vm/issues/1122
+    push.1 drop adv.push_mapval
+    # => [TYPES_COM]
+
+    # get the storage slot types data pointer
+    exec.layout::get_acct_storage_slot_type_data_ptr
+    # => [slot_type_data_ptr, TYPES_COM]
+
+    # prepare the stack for reading storage slot types
+    padw padw padw
+    # => [ZERO, ZERO, ZERO, slot_type_data_ptr, TYPES_COM]
+
+    # read the storage slot types from the advice provider and store in memory
+    repeat.32
+        adv_pipe hperm
+        # => [PERM, PERM, PERM, slot_type_data_ptr', TYPES_COM]
+    end
+
+    # extract the digest and drop the memory pointer
+    dropw swapw dropw movup.4 drop
+    # => [DIGEST, TYPES_COM]
+
+    # assert the digest matches the commitment
+    assert_eqw
+    # => []
+end
+
+#! Validates that storage slot types are well formed.
+#!
+#! This is achieved by reading the storage slot types from memory and asserting that they are
+#! well formed. Currently we only support scalar slots with an entry of maximum arity of 256.
+#!
+#! Stack: []
+#! Output: []
+proc.validate_storage_slot_types
+    # get the pointer at which the storage slot type data begins
+    exec.layout::get_acct_storage_slot_type_data_ptr
+    # => [slot_type_data_ptr]
+
+    repeat.32
+        # read the storage slot types from memory
+        padw dup.4 mem_loadw
+        # => [slot_i_type, slot_i+1_type, slot_i+2_type, slot_i+3_type, slot_type_data_ptr]
+
+        repeat.4
+            # split the type to get the slot type and entry arity
+            u32split
+            # => [arity, type, ...]
+
+            # assert the entry arity is less than or equal to 255
+            push.255 lte assert
+            # => [type, ...]
+
+            # assert the slot type is valid
+            exec.account::get_max_slot_type lte assert
+            # => [...]
```

Comment:
We could make these a bit more efficient by replacing `push.255 lte assert` with something like this:
```
u32split assertz push.255 u32lte assert
```
And `exec.account::get_max_slot_type lte assert` with something like this:
```
u32split assertz exec.account::get_max_slot_type u32lte assert
```
If this works, it should reduce cycle count by almost 50%.